### DISCUSSION TITLE
I want to draw mixed font family textline, and get 4-point coordinates for each character drawed, how to realize it with pillow?

### DISCUSSION THREAD

**Discussion by nissansz** [python-pillow/Pillow#8305] [Q&A]
I want to draw mixed font family textline, and get 4-point coordinates for each character drawed, how to realize it with pillow?

**Comment by radarhere** [comment]
To be clear, I'm guessing that a 'textline' is just a single line of text.

To draw text with different fonts, I suggest you just draw each character individually.
To get the co-ordinates, I suggest you call [`textbbox()`](https://pillow.readthedocs.io/en/stable/reference/ImageDraw.html#PIL.ImageDraw.ImageDraw.textbbox) for each individual character.

**Comment by nissansz** [comment]
Yes, a single line of text.
mixed with multiple fonts/languages, can you help share a full code?

**Comment by nissansz** [comment]
bbox = draw.textbbox((current_x, current_y), char, font=font)
            char_width = bbox[2] - bbox[0]  
            char_height = bbox[3] - bbox[1]  
bbox seems not correct somtimes, how to improve? There is a large blank gap sometimes

![image](https://github.com/user-attachments/assets/0cf02c06-5724-4b01-9f00-46579c7214a8)

**Reply by radarhere** [reply]
Are you making use of libraqm? That could improve the results.
```python
ImageFont.truetype(..., layout_engine=ImageFont.Layout.RAQM)
```

**Reply by nissansz** [reply]
UserWarning: Raqm layout was requested, but Raqm is not available. Falling back to basic layout.
  warnings.warn(  
  
  How to install it?

**Reply by radarhere** [reply]
See https://github.com/python-pillow/Pillow/issues/7337 for where you have asked about installing raqm before.
